Add oclif v2 announcement blog (#130)
oclif/oclif.github.io: website/blog/2022-01-12-announcing-oclif-v2.md

* `@oclif/command`
* `@oclif/config`
* `@oclif/error`
* `@oclif/help`
* `@oclif/parser`

### No more single vs multi command CLIs

There is only 1 "type" of oclif v2 CLI but it can have as few or as many commands as a developer wants. This also means there’s only one command for generating CLIs, `oclif generate ` as opposed to the old `oclif single` and `oclif multi` commands.

### Node > 12

In order to ensure oclif CLIs are as secure as possible all v2 libraries and plugins only support Node 12 or higher as of now and, going forward, will only be supporting Node Maintenance or higher as defined by Node's release schedule here: https://nodejs.org/en/about/releases/.

## What’s new?

### New example CLI

[`oclif-hello-world`](https://github.com/oclif/hello-world/) is our new example repo. This is also the repo that is [used as a template](https://github.com/oclif/oclif/blob/edc6616e51/src/generators/cli.ts#L74) when running `oclif generate` to create a new CLI. This repo will always be able to be referenced as an example of what an oclif v2 CLI should look like.

### Spaced commands

> oclif CLIs can now use spaces to separate topics and subcommands instead of colons. To enable this feature simply add `“topicSeparator": " "` to the oclif config in your package.json. You can see an example of this in our [example repo](https://github.com/oclif/hello-world/blob/main/package.json#L55).

Reviewer: You might want to add that colons will continue to work even if spaces are specified, which means that people can enable spaces without do a major version bump or breaking their customers

Author: Oh yes, good point!

### Async Command Parsing

We’ve also improved the performance of new CLIs by rewriting how commands are parsed on initialization. The [new parser](https://github.com/oclif/core/blob/main/src/parser/parse.ts) is asynchronous which makes CLIs with a lot of commands or installed  plugins much faster.

## What’s coming?

Part of our charter for the release of oclif v2 includes improving our engagement with the oclif community. We know that over the past couple of years we’ve reduced our involvement and a lot of issues and PRs have languished in limbo. Hopefully you’ve already seen increased activity and responses from oclif maintainers recently but if you haven’t please don’t hesitate to ping us by tagging [@admins](https://github.com/orgs/oclif/teams/admins) in oclif repos.

We also want to start interviewing members of the oclif community to acquire feedback and prioritize the features and fixes you deem most important!

Best,

The oclif team

## Reference material 

### Migration Guide

> This guide explains how to upgrade a CLI or plugin from the old Ioclif v1 libraries to the new `@oclif/core` library oclif v2 uses.

Reviewer: typo: `Ioclif` -> `oclif`

---
title: Announcing oclif v2!
---

> Hello and happy new year! It is our great pleasure to announce that with the turn of a new year we have finally released oclif v2 which uses the new [@oclif/core](https://github.com/oclif/core) library! 🎉

Reviewer 2: ``` suggestion
Hello and happy new year! It is our great pleasure to announce that with the turn of a new year we have finally released oclif v2, which uses the new [@oclif/core](https://github.com/oclif/core) library! 🎉
```

## What’s changing?

### Repository Changes

All repositories under the [oclif org](https://github.com/oclif/) now use `main` as their primary branch and `oclif-v1` as the legacy branch. The `main`  branch corresponds to oclif v2 and the `oclif-v1` branch to oclif v1. If you find any repos that don’t have a `main` branch you can safely assume that they are not used in oclif v2 and will be deprecated at some point in the future.

### Consolidating Tools & Libraries

The [`oclif-dev`](https://github.com/oclif/dev-cli) CLI has now been incorporated into [`oclif`](https://github.com/oclif/oclif). Now you no longer need to install a separate package to manage your entire CLI’s lifecycle.

The following libraries have been consolidated under [`@oclif/core`](https://github.com/oclif/core) and will be deprecated at some point in the future. You can read the [migration guide](https://github.com/oclif/core/blob/main/MIGRATION.md) to see how to update your CLI or plugin to use the new core library.

* `@oclif/command`
* `@oclif/config`
* `@oclif/error`
* `@oclif/help`
* `@oclif/parser`

### No more single vs multi command CLIs

> There is only 1 "type" of oclif v2 CLI but it can have as few or as many commands as a developer wants. This also means there’s only one command for generating CLIs, `oclif generate ` as opposed to the old `oclif single` and `oclif multi` commands.

Reviewer 2: Aren't these being restored?

Author: Nope.

### Node > 12

In order to ensure oclif CLIs are as secure as possible all v2 libraries and plugins only support Node 12 or higher as of now and, going forward, will only be supporting Node Maintenance or higher as defined by Node's release schedule here: https://nodejs.org/en/about/releases/.

> 

Reviewer 2: ``` suggestion
In order to ensure oclif CLIs are as secure as possible all v2 libraries and plugins only support Node 12 or higher as of now and, going forward, will only be supporting Node Maintenance or higher as defined by Node's [release schedule](https://nodejs.org/en/about/releases/).
```

Author: I kept this as a separate URL and not a link on purpose to make sure people look that we're linking directly to nodejs.org.

---
title: Announcing oclif v2!
---

> Hello and happy new year! It is our great pleasure to announce that with the turn of a new year we have finally released oclif v2, which uses the new [@oclif/core](https://github.com/oclif/core) library! 🎉

Reviewer 3: ```suggestion
Hello and happy new year! It's our great pleasure to announce that we have released oclif v2, which uses the new [@oclif/core](https://github.com/oclif/core) library! 🎉
```

oclif v2 includes many changes that improve the experience of both creating and using an oclif CLI.

> Going forward we do not plan to make any updates to oclif v1 and its corresponding libraries except for critical security fixes (please see the [compatibility matrix](#Compatibility-Matrix) for a list of these libraries). We are not completely dropping support yet in order to give developers time to migrate from v1 to v2 but at some point in the future we will archive v1 repositories and deprecate their versions on NPM.

Reviewer 3: ```suggestion
Going forward, we don't plan to make any updates to oclif v1 and its corresponding libraries, except for critical security fixes. See the [compatibility matrix](#Compatibility-Matrix) for a list of these libraries. In order to give developers time to migrate from v1 to v2, we're not completely dropping support yet. But at some point in the future we'll archive the v1 repositories and deprecate their versions on NPM.
```

## What’s changing?

### Repository Changes

> All repositories under the [oclif org](https://github.com/oclif/) now use `main` as their primary branch and `oclif-v1` as the legacy branch. The `main`  branch corresponds to oclif v2 and the `oclif-v1` branch to oclif v1. If you find any repos that don’t have a `main` branch you can safely assume that they are not used in oclif v2 and will be deprecated at some point in the future.

Reviewer 3: ```suggestion
All repositories under the [oclif org](https://github.com/oclif/) now use `main` as their primary branch and `oclif-v1` as the legacy branch. The `main`  branch corresponds to oclif v2 and the `oclif-v1` branch to oclif v1. If you find any repos that don’t have a `main` branch, you can safely assume that they're not used in oclif v2 and will be deprecated in the future.
```

### Consolidating Tools & Libraries

> The [`oclif-dev`](https://github.com/oclif/dev-cli) CLI has now been incorporated into [`oclif`](https://github.com/oclif/oclif). Now you no longer need to install a separate package to manage your entire CLI’s lifecycle.

Reviewer 3: ```suggestion
The [`oclif-dev`](https://github.com/oclif/dev-cli) CLI has now been incorporated into [`oclif`](https://github.com/oclif/oclif). You now no longer need to install a separate package to manage your entire CLI’s lifecycle.
```

> The following libraries have been consolidated under [`@oclif/core`](https://github.com/oclif/core) and will be deprecated at some point in the future. You can read the [migration guide](https://github.com/oclif/core/blob/main/MIGRATION.md) to see how to update your CLI or plugin to use the new core library.

Reviewer 3: ```suggestion
The following libraries have been consolidated under [`@oclif/core`](https://github.com/oclif/core) and will be deprecated at some point in the future. Read the [migration guide](https://github.com/oclif/core/blob/main/MIGRATION.md) to learn how to update your CLI or plugin to use the new core library.
```

* `@oclif/command`
* `@oclif/config`
* `@oclif/error`
* `@oclif/help`
* `@oclif/parser`

### No more single vs multi command CLIs

> There is only 1 "type" of oclif v2 CLI but it can have as few or as many commands as a developer wants. This also means there’s only one command for generating CLIs, `oclif generate` as opposed to the old `oclif single` and `oclif multi` commands.

Reviewer 3: ```suggestion
There's only one "type" of oclif v2 CLI, but it can have as few or as many commands as a developer wants. As a result, there’s only one command to generate CLIs, `oclif generate`, as opposed to the old `oclif single` and `oclif multi` commands.
```

### Node > 12

> In order to ensure oclif CLIs are as secure as possible all v2 libraries and plugins only support Node 12 or higher as of now and, going forward, will only be supporting Node Maintenance or higher as defined by Node's release schedule here: https://nodejs.org/en/about/releases/.

Reviewer 3: ```suggestion
To ensure oclif CLIs are as secure as possible, all v2 libraries and plugins support only Node 12 or higher as of now. Going forward, they'll support only Node Maintenance or higher, as defined by Node's release schedule here: https://nodejs.org/en/about/releases/.
```

Reviewer 3: Okay?  I hope I didn't change the meaning with my edit.  Please read carefully :)

## What’s new?

### New example CLI

> [`oclif-hello-world`](https://github.com/oclif/hello-world/) is our new example repo. This is also the repo that is [used as a template](https://github.com/oclif/oclif/blob/edc6616e51/src/generators/cli.ts#L74) when running `oclif generate` to create a new CLI. This repo will always be able to be referenced as an example of what an oclif v2 CLI should look like.

Reviewer 3: ```suggestion
[`oclif-hello-world`](https://github.com/oclif/hello-world/) is our new example repo. This is also the repo that's [used as a template](https://github.com/oclif/oclif/blob/edc6616e51/src/generators/cli.ts#L74) when running `oclif generate` to create a new CLI. This repo will always be able to be referenced as an example of what an oclif v2 CLI should look like.
```

### Spaced commands

> oclif CLIs can now use spaces to separate topics and subcommands instead of colons. To enable this feature simply add `“topicSeparator": " "` to the oclif config in your package.json. You can see an example of this in our [example repo](https://github.com/oclif/hello-world/blob/main/package.json#L55).

Reviewer 3: ```suggestion
oclif CLIs can now use spaces, instead of colons, to separate topics and subcommands. To enable this feature, simply add `“topicSeparator": " "` to the oclif config in your package.json. See an example in our [example repo](https://github.com/oclif/hello-world/blob/main/package.json#L55).
```

```
// Old commands that use :
$ mycli do:something
```
```
// New spaced commands
$ mycli do something
```

> > Note: Spaced commands are backwards compatible so if you enable spaced commands for your CLI users will still be able to use `:` as a separator. This ensures that any existing scripts are not broken.

Reviewer 3: ```suggestion
> Note: Spaced commands are backwards compatible. So if you enable spaced commands for your CLI, users will still be able to use `:` as a separator. This ensures that any existing scripts don't break.
```

### New Help Output

> We’ve revamped the way help for commands is output to the terminal to make it both easier to implement and easier to read. You can see the difference between the old help output on the left and the new help output on the right in the screenshots down below.

Reviewer 3: ```suggestion
We’ve revamped the way command help is outputted to the terminal, making it both easier to implement and easier to read. See the difference between the old help output on the left and the new help output on the right in the screenshots below.
```

<table border="0">
 <tr>
    <td>oclif v1</td>
    <td>oclif v2</td>
 </tr>
 <tr>
    <td><img src="/img/2022-01-12-announcing-oclif-v2/sfdx-help.png"/></td>
    <td><img src="/img/2022-01-12-announcing-oclif-v2/sf-help.png"/></td>
 </tr>
</table>

> You’ll notice that there are new sections for flags and global flags, examples are displayed with better spacing, and there is a section at the bottom called Configuration Variables. The Configuration Variables section is not part of the new help by default but we now have support for custom help sections which is what the `sf` CLI uses to create it.

Reviewer 3: ```suggestion
Notice that there are new sections for flags and global flags, examples are displayed with better spacing, and there is a section at the bottom called Configuration Variables. This Configuration Variables section is not part of the new help by default. But we've added support for custom help sections, which is what the `sf` CLI uses to create the new section. 
```

### Async Command Parsing

> We’ve also improved the performance of new CLIs by rewriting how commands are parsed on initialization. The [new parser](https://github.com/oclif/core/blob/main/src/parser/parse.ts) is asynchronous which makes CLIs with a lot of commands or installed  plugins much faster.

Reviewer 3: ```suggestion
We’ve also improved the performance of new CLIs by rewriting how commands are parsed on initialization. The [new parser](https://github.com/oclif/core/blob/main/src/parser/parse.ts) is asynchronous, which makes CLIs with a lot of commands or installed  plugins much faster.
```

> ## What’s coming?

Reviewer 3: ```suggestion
## What’s coming next?
```

> Part of our charter for the release of oclif v2 includes improving our engagement with the oclif community. We know that over the past couple of years we’ve reduced our involvement and a lot of issues and PRs have languished in limbo. Hopefully you’ve already seen increased activity and responses from oclif maintainers recently but if you haven’t please don’t hesitate to ping us by tagging [@admins](https://github.com/orgs/oclif/teams/admins) in oclif repos.

Reviewer 3: ```suggestion
Part of our charter for the release of oclif v2 includes improving our engagement with the oclif community. We know that over the past couple of years we’ve reduced our involvement, and a lot of issues and PRs have languished in limbo. Hopefully you’ve already seen increased activity and responses from oclif maintainers recently, but if you haven’t, please don’t hesitate to ping us by tagging [@admins](https://github.com/orgs/oclif/teams/admins) in oclif repos.
```

Reviewer 3: "languishing in limbo" -- love it!

We also want to start interviewing members of the oclif community to acquire feedback and prioritize the features and fixes you deem most important!

Best,

The oclif team

<br/>

### Reference material 

#### Migration Guide

> This guide explains how to upgrade a CLI or plugin from the old oclif v1 libraries to the new `@oclif/core` library oclif v2 uses.

Reviewer 3: ```suggestion
This guide explains how to upgrade a CLI or plugin from the old oclif v1 libraries to the new `@oclif/core` library that oclif v2 uses.
```

https://github.com/oclif/core/blob/main/MIGRATION.md

#### Compatibility Matrix

> The following matrix shows how the v1 libraries and plugins relate to the new ones for v2. You can use this as a guide to know what to drop and which versions to switch when upgrading your plugins and CLIs to v2.

Reviewer 3: ```suggestion
The following matrix shows how the v1 libraries and plugins relate to the new v2 ones. Use this matrix as a guide to know what to drop and which versions to switch when upgrading your plugins and CLIs to v2.
```
